File: python/ch06/two_opt.py
# python/ch06/two_opt.py
"""TSP 玩具: ランダム初期巡回 + 2-opt."""
from __future__ import annotations

import math
import random
import sys
from typing import List, TextIO, Tuple


def read_points(stream: TextIO) -> List[Tuple[float, float]]:
    lines = stream.read().splitlines()
    n = int(lines[0])
    pts: List[Tuple[float, float]] = []
    for i in range(1, n + 1):
        x, y = map(float, lines[i].split())
        pts.append((x, y))
    return pts


def tour_length(pts: List[Tuple[float, float]], tour: List[int]) -> float:
    total = 0.0
    for i in range(len(tour)):
        a = pts[tour[i]]
        b = pts[tour[(i + 1) % len(tour)]]
        total += math.hypot(a[0] - b[0], a[1] - b[1])
    return total


def two_opt(pts: List[Tuple[float, float]], tour: List[int]) -> List[int]:
    n = len(tour)
    improved = True
    while improved:
        improved = False
        for i in range(n):
            for k in range(i + 2, n if i > 0 else n - 1):
                a, b = tour[i], tour[(i + 1) % n]
                c, d = tour[k], tour[(k + 1) % n]
                before = (
                    math.hypot(pts[a][0] - pts[b][0], pts[a][1] - pts[b][1])
                    + math.hypot(pts[c][0] - pts[d][0], pts[c][1] - pts[d][1])
                )
                after = (
                    math.hypot(pts[a][0] - pts[c][0], pts[a][1] - pts[c][1])
                    + math.hypot(pts[b][0] - pts[d][0], pts[b][1] - pts[d][1])
                )
                if after + 1e-9 < before:
                    tour[i + 1 : k + 1] = reversed(tour[i + 1 : k + 1])
                    improved = True
    return tour


def main() -> None:
    pts = read_points(sys.stdin)
    rng = random.Random(42)
    tour = list(range(len(pts)))
    rng.shuffle(tour)
    before = tour_length(pts, tour)
    tour = two_opt(pts, tour)
    after = tour_length(pts, tour)
    print("tour:", " ".join(str(i) for i in tour))
    print(f"n={len(pts)} before={before:.2f} after={after:.2f}")


if __name__ == "__main__":
    main()

File: go/ch06/two_opt.go
// go/ch06/two_opt.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

type Point struct {
	X, Y float64
}

func readPoints(sc *bufio.Scanner) ([]Point, error) {
	var lines []string
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}
	if err := sc.Err(); err != nil {
		return nil, err
	}
	n, _ := strconv.Atoi(lines[0])
	pts := make([]Point, n)
	for i := 0; i < n; i++ {
		p := strings.Fields(lines[i+1])
		x, _ := strconv.ParseFloat(p[0], 64)
		y, _ := strconv.ParseFloat(p[1], 64)
		pts[i] = Point{x, y}
	}
	return pts, nil
}

func tourLength(pts []Point, tour []int) float64 {
	total := 0.0
	n := len(tour)
	for i := 0; i < n; i++ {
		a := pts[tour[i]]
		b := pts[tour[(i+1)%n]]
		total += math.Hypot(a.X-b.X, a.Y-b.Y)
	}
	return total
}

func edgeDist(pts []Point, a, b int) float64 {
	return math.Hypot(pts[a].X-pts[b].X, pts[a].Y-pts[b].Y)
}

func twoOpt(pts []Point, tour []int) []int {
	n := len(tour)
	improved := true
	for improved {
		improved = false
		for i := 0; i < n; i++ {
			limit := n
			if i == 0 {
				limit = n - 1
			}
			for k := i + 2; k < limit; k++ {
				a, b := tour[i], tour[(i+1)%n]
				c, d := tour[k], tour[(k+1)%n]
				before := edgeDist(pts, a, b) + edgeDist(pts, c, d)
				after := edgeDist(pts, a, c) + edgeDist(pts, b, d)
				if after+1e-9 < before {
					reverseSegment(tour, i+1, k)
					improved = true
				}
			}
		}
	}
	return tour
}

func reverseSegment(tour []int, lo, hi int) {
	for lo < hi {
		tour[lo], tour[hi] = tour[hi], tour[lo]
		lo++
		hi--
	}
}

func main() {
	pts, err := readPoints(bufio.NewScanner(os.Stdin))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	rng := rand.New(rand.NewSource(42))
	tour := make([]int, len(pts))
	for i := range tour {
		tour[i] = i
	}
	rng.Shuffle(len(tour), func(i, j int) { tour[i], tour[j] = tour[j], tour[i] })
	before := tourLength(pts, tour)
	tour = twoOpt(pts, tour)
	after := tourLength(pts, tour)
	fmt.Print("tour:")
	for _, i := range tour {
		fmt.Printf(" %d", i)
	}
	fmt.Println()
	fmt.Printf("n=%d before=%.2f after=%.2f\n", len(pts), before, after)
}
